common/src/main/java/org/vivecraft/server/config/ConfigBuilder.java

```diff
@@ -225,32 +228,111 @@ public T reset() {
+    }
+
+    public static class ListValue<T> extends ConfigValue<List<T>>{
+        public ListValue(CommentedConfig config, List<String> path, List<T> defaultValue) {
             super(config, path, defaultValue);
         }
     }
 
-    public static class IntValue extends ConfigValue<Integer>{
-        public IntValue(Config config, List<String> path, int defaultValue) {
+    public static class InListValue<T> extends ConfigValue<T>{
+        private final Collection<? extends T> validValues;
+        public InListValue(CommentedConfig config, List<String> path, T defaultValue, Collection<? extends T> validValues) {
+            super(config, path, defaultValue);
+            this.validValues = validValues;
+        }
+
+        public Collection<? extends T> getValidValues(){
+            return validValues;
+        }
+    }
+
+    public interface NumberValue<E extends Number>{
+        E getMin();
+        E getMax();
+        E get();
+        E reset();
+        void set(E newValue);
+        double normalize();
+        void fromNormalized(double value);
+    }
+
+    public static class IntValue extends ConfigValue<Integer> implements NumberValue<Integer> {
+
+        private final int min;
+        private final int max;
+
+        public IntValue(CommentedConfig config, List<String> path, int defaultValue, int min, int max) {
             super(config, path, defaultValue);
+            this.min = min;
+            this.max = max;
+        }
+
+        public Integer getMin() {
+            return min;
+        }
+
+        public Integer getMax() {
+            return max;
+        }
+
+        public double normalize() {
+            return Mth.clamp((this.get() - min) / (double) (max - min), 0.0D, 1.0D);
+        }
+
+        public void fromNormalized(double value) {
+            double newValue = this.getMin() + (this.getMax() - this.getMin()) * value;
+            this.set(Mth.floor(newValue + 0.5));
         }
     }
 
-    public static class DoubleValue extends ConfigValue<Double>{
-        public DoubleValue(Config config, List<String> path, double defaultValue) {
+    public static class DoubleValue extends ConfigValue<Double> implements NumberValue<Double> {
+
+        private final double min;
+        private final double max;
+        public DoubleValue(CommentedConfig config, List<String> path, double defaultValue, double min, double max) {
             super(config, path, defaultValue);
+            this.min = min;
+            this.max = max;
+        }
+        public Double getMin() {
+            return min;
+        }
+        public Double getMax() {
+            return max;
+        }
+        public double normalize() {
+            return Mth.clamp((this.get() - min) / (max - min), 0.0D, 1.0D);
+        }
+
+        public void fromNormalized(double value) {
+            double newValue = this.getMin() + (this.getMax() - this.getMin()) * value;
+            BigDecimal bd = BigDecimal.valueOf(newValue);
```

Comment:
Not sure we need to use big decimal here for rounding

Reply:
is there another way to round to 2 decimal places?
could use some integer math probably

Reply:
Math.round(value *100) / 100.0 works as well iirc.

Reply:
hm, okay

Reply:
fixed